### PR TITLE
feat(android): chunked video recording for API <34 and infinite recording for API >=34

### PR DESCRIPTION
**Pull request by Malinskiy**

### CODE REVIEW

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/AdamScreenCaptureTestRunListener.kt [hunk 1]
```diff
@@ -56,7 +56,7 @@ class AdamScreenCaptureTestRunListener(
                 device.safePullFile(path, localFile.absolutePath)
                 logger.debug { "Received screen capture file $path" }
                 attachmentListeners.forEach {
-                    it.onAttachment(test.toTest(), Attachment(localFile, attachmentType))
+                    it.onAttachment(test.toTest(), Attachment(localFile, attachmentType, "screen"))
```
Comment (matzuk) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/AdamScreenCaptureTestRunListener.kt]: `screen` should be a const or enum?

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt [hunk 2]
```diff
@@ -32,7 +31,65 @@ internal class ScreenRecorder(
         logger.debug { "Recording finished in ${millis}ms $remoteFilePath" }
     }
 
+    suspend fun stopScreenRecord() {
+        logger.debug { "Stopping screen recorder" }
+        var hasKilledScreenRecord = true
+        var tries = 0
+        while (hasKilledScreenRecord && tries++ < SCREEN_RECORD_KILL_ATTEMPTS) {
```
Comment (matzuk) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt]: I recommend keeping the rule "one line - one action."
Could you put `tries++` in a separate line?

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt [hunk 3]
```diff
@@ -35,54 +36,78 @@ class ScreenRecorderTestBatchListener(
...
+    private val screenRecorder = ScreenRecorder(device, videoConfiguration)
 
     private var hasFailed: Boolean = false
     private var supervisorJob: Job? = null
-    private var lastRemoteFile: String? = null
+    private var lastRemoteFiles: MutableList<String> = mutableListOf()
+    private val chunkedRecording = device.apiLevel < 34
+    private var chunkMap: HashMap<TestIdentifier, Long> = HashMap<TestIdentifier, Long>().apply { withDefault { 0 } }
+    @Volatile
+    private var cancelling: Boolean = false
```
Comment (matzuk) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt:55]: `@volatile` is enough? Maybe `AtomicBoolean`?

File: vendor/vendor-android/src/test/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListenerTest.kt [hunk 4]
```diff
@@ -64,8 +65,9 @@ class ScreenRecorderTestRunListenerTest {
                 serial("emulator-5554") {
                     boot()
                     shell(
-                        "screenrecord --size 720x1280 --bit-rate 1000000 --time-limit 180 /sdcard/com.example.Class-method1-${batch.id}.mp4",
-                        ""
+                        "screenrecord --size 720x1280 --bit-rate 1000000 --time-limit 180 /sdcard/com.example.Class-method1-${batch.id}-0.mp4",
```
Comment (matzuk) [inline vendor/vendor-android/src/test/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListenerTest.kt:68]: Is it possible to write tests checking that two or more chunks were created?

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/AdamScreenCaptureTestRunListener.kt (hunk 1 above)
Comment (Malinskiy) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/AdamScreenCaptureTestRunListener.kt]: every vendor can choose what to use as a name for the attachment. I doubt we should restrict anything here, but I can collect these values under an enum with optional overrides. For example log is named xcodebuild-log for apple and logcat for Android

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt (hunk 2 above)
Comment (Malinskiy) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt]: Not a `++` fan I assume

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt (hunk 3 above)
Comment (Malinskiy) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt:55]: It should be enough for the purpose of synchronizing the stop of the screen recorder

File: vendor/vendor-android/src/test/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListenerTest.kt (hunk 4 above)
Comment (Malinskiy) [reply vendor/vendor-android/src/test/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListenerTest.kt:68]: In theory yes, but you'd have to wait more. Scheduler is not mocked here.

**Review by matzuk — APPROVED** [review]

**Pull request merged**